### PR TITLE
client: allow reverting finalized blocks

### PR DESCRIPTION
**Pull request by andresilva**
This PR extends the revert functionality to allow reverting past finalized blocks. It also fixes the handling of leaves when reverting blocks.

### CODE REVIEW

File: client/db/src/lib.rs [hunk 1]
```diff
@@ -1484,40 +1484,71 @@ impl<Block> sc_client_api::backend::Backend<Block, Blake2Hasher> for Backend<Blo
...
-						|| sp_blockchain::Error::UnknownBlock(
-							format!("Error reverting to {}. Block hash not found.", best)))?;
-					let key = utils::number_and_hash_to_lookup_key(best.clone(), &hash)?;
-					transaction.put(columns::META, meta_keys::BEST_BLOCK, &key);
-					transaction.delete(columns::KEY_LOOKUP, removed.hash().as_ref());
-					children::remove_children(&mut transaction, columns::META, meta_keys::CHILDREN_PREFIX, hash);
-					self.storage.db.write(transaction).map_err(db_err)?;
-					self.blockchain.update_meta(hash, best, true, false);
-					self.blockchain.leaves.write().revert(removed.hash().clone(), removed.number().clone(), removed.parent_hash().clone());
+		let revertible = best_number - finalized;
```
Comment (bkchr) [inline client/db/src/lib.rs]:
```suggestion
```

File: client/db/src/lib.rs [hunk 2]
```diff
@@ -1484,40 +1484,71 @@ impl<Block> sc_client_api::backend::Backend<Block, Blake2Hasher> for Backend<Blo
...
+		let n = if !revert_finalized && revertible < n {
```
Comment (bkchr) [inline client/db/src/lib.rs]:
```suggestion
		let n = if !revert_finalized {
```

File: client/db/src/lib.rs [hunk 3]
```diff
@@ -1484,40 +1484,71 @@ impl<Block> sc_client_api::backend::Backend<Block, Blake2Hasher> for Backend<Blo
...
+			revertible
```
Comment (bkchr) [inline client/db/src/lib.rs]:
```suggestion
			best_number - finalized
```

File: client/db/src/lib.rs (hunk 2 above)
Comment (gavofyork) [reply client/db/src/lib.rs]: this change would alter the logic. explanation?

File: client/db/src/lib.rs [hunk 4]
```diff
@@ -1484,40 +1484,71 @@ impl<Block> sc_client_api::backend::Backend<Block, Blake2Hasher> for Backend<Blo
...
+			n
+		};
+
+		let mut revert_blocks = || -> ClientResult<NumberFor<Block>> {
+			for c in 0 .. n.saturated_into::<u64>() {
+				if best_number.is_zero() {
+					return Ok(c.saturated_into::<NumberFor<Block>>())
+				}
+				let mut transaction = DBTransaction::new();
+				match self.storage.state_db.revert_one() {
```
Comment (arkpar) [inline client/db/src/lib.rs]: I believe this still won't revert past the last canonicalized block when pruning is enabled. `revert_one` returns `None` for finalized blocks.

Comment (andresilva) [reply client/db/src/lib.rs]: Yes, this only works with `pruning=archive` (i.e. for validators). It will stop reverting if it has no state.

**Comment by kigawas** [comment]
In which case it'll need to revert finalized blocks?

**Comment by andresilva** [comment]
@kigawas https://medium.com/@gavofyork/kusamas-first-adventure-2cd4f439a7a4 A writeup of why we needed this for Kusama.

**Pull request merged**